homeassistant/components/calendar/todoist.py

```diff
@@ -0,0 +1,595 @@
+
+    def cleanup(self):
+        """Clean up all calendar data."""
+        super().cleanup()
+        self._cal_data['all_tasks'] = []
+
+    @property
+    def device_state_attributes(self):
+        """Return the device state attributes."""
+        if self.data.event is None:
+            # No tasks, we don't REALLY need to show anything.
+            return {}
+
+        attributes = super().device_state_attributes
+
+        # Add additional attributes.
+        attributes['due_today'] = self.data.event['due_today']
+        attributes['overdue'] = self.data.event['overdue']
+        attributes['all_tasks'] = self._cal_data['all_tasks']
+        attributes['priority'] = self.data.event['priority']
+        attributes['task_labels'] = self.data.event['labels']
+
+        return attributes
+
+
+class TodoistProjectData(object):
+    """
+    Class used by the Task Device service object to hold all Todoist Tasks.
+
+    This is analagous to the GoogleCalendarData found in the Google Calendar
+    component.
+
+    Takes an object with a 'name' field and optionally an 'id' field (either
+    user-defined or from the Todoist API), a Todoist API token, and an optional
+    integer specifying the latest number of days from now a task can be due (7
+    means everything due in the next week, 0 means today, etc.).
+
+    This object has an exposed 'event' property (used by the Calendar platform
+    to determine the next calendar event) and an exposed 'update' method (used
+    by the Calendar platform to poll for new calendar events).
+
+    The 'event' is a representation of a Todoist Task, with defined parameters
+    of 'due_today' (is the task due today?), 'all_day' (does the task have a
+    due date?), 'task_labels' (all labels assigned to the task), 'message'
+    (the content of the task, e.g. 'Fetch Mail'), 'description' (a URL pointing
+    to the task on the Todoist website), 'end_time' (what time the event is
+    due), 'start_time' (what time this event was last updated), 'overdue' (is
+    the task past its due date?), 'priority' (1-4, how important the task is,
+    with 4 being the most important), and 'all_tasks' (all tasks in this
+    project, sorted by how important they are).
+
+    'offset_reached', 'location', and 'friendly_name' are defined by the
+    platform itself, but are not used by this component at all.
+
+    The 'update' method polls the Todoist API for new projects/tasks, as well
+    as any updates to current projects/tasks. This is throttled to every
+    MIN_TIME_BETWEEN_UPDATES minutes.
+    """
+
+    def __init__(self, project_data, labels, token,
+                 latest_task_due_date=None, whitelisted_labels=None,
+                 whitelisted_projects=None):
+        """Initialize a Todoist Project."""
+        self.event = None
+
+        self._token = token
+        self._name = project_data['name']
+        # If no ID is defined, fetch all tasks.
+        if project_data['id'] is not None:
```

Comment:
So as it turns out, that's not a dict: it's a `Project` object from the Todoist API. I had been operating under the assumption that it was a dict, [since that's what their API seems to imply.](https://developer.todoist.com/sync/v7/#projects) I'm not sure where exactly that `Project` object is stored in their API; `from todoist import Project` doesn't seem to work, and they don't really have it documented much in their API. I can create a new project using their API calls, but then it'd be synchronized with their server.

All I *really* need from their `Project` object is the name and ID of the project, so I can either handle it with a `try-except` block here to catch the missing key or just have the method accept a flat-out dict instead of a `Project` object. I'll likely just go with the latter, since it seems less hacky.